src/realm/exceptions.hpp

```diff
@@ -201,7 +229,7 @@ class DuplicatePrimaryKeyValueException : public std::logic_error {
 /// exception being thrown. The whole point of properly documenting "Undefined
 /// Behaviour" cases is to help the user know what the limits are, without
 /// constraining the database to handle every and any use-case thrown at it.
-class LogicError : public ExceptionWithBacktrace<std::exception> {
+class LogicError : public ExceptionForStatus {
```

Comment:
Having to parse exception messages to read the things we care about would be a huge step backwards.

Reply:
...did someone suggest having to parse exception messages? I certainly didn't.

Reply:
We currently read data members from exception subclasses and you're saying we shouldn't have those, so yes?

Right now the big awkward thing about core exceptions is that most of them don't have messages which we can just forward on, and don't provide enough information to construct good error messages or populate the userInfo dictionary. We currently work around this by duplicating all of the parameter validation done by core so that we don't rely on the exceptions being useful. Removing data members from the exceptions which do have them and only having an error code and message isn't really a step in the right direction?

Reply:
@tgoyne is there a different way we can approach this then? Part of the point of exception unification project is so Core just has a single exception type.

Reply:
Well I guess my question is why is that the point? That seems like something that might be useful as a means to an end (e.g. to simplify converting between exceptions and errors) but not as a goal on its own. The scope doc claims it'd make it easier to catch exceptions thrown by core, which doesn't make any sense. The scope doc also says that the new error type should have more context, not less.

Reply:
I guess it would be easier from the perspective of only having to catch a single error type in cpp as opposed to knowing what errors can be thrown.

But in reading your comments, it seems like the more important improvement here for us would just be for core to provide the correct error message so we don't have to do that at the SDK level.

Reply:
For the catch-and-forward use case there's no difference between catching Exception when that's the only exception type defined and catching Exception when there's also subclasses. Doing anything more complicated inherently requires knowing what types of errors may occur regardless of how those are represented.

The difficult part of throwing exceptions with usable error messages from core is that the desired error message depends on the language and SDK. For example, take [RLMThrowResultsError](https://github.com/realm/realm-swift/blob/master/Realm/RLMResults.mm#L79). The exception message in `OutOfBoundsIndexException` is fine, but it's a different message from what out-of-bounds access to NSArray has, and we want to match the platform as much as possible. `UnsupportedColumnTypeException` has a pretty reasonable message, but it says that `max()` doesn't support the property type rather than `maximumOfProperty:`.

Our original idea for how to handle this in object store which we never got around to implementing was to consolidate all exception message construction to a single place, and then have per-SDK/language variations on that single place.

Outside of constructing error messages, the primary bit of information which we grab from exceptions that is not covered by an error code is the file path for filesystem-related errors, which we store in the userInfo dictionary. Anything which involves a call to a system function is also generally going to have an underlying error code which we want to preserve in addition to our error code.